### PR TITLE
Respect EOL character from `.editorconfig`

### PR DESCRIPTION
**Pull request by CloudNStoyan**
Fixes https://github.com/bmish/eslint-doc-generator/issues/527

It uses the `editorconfig` npm package https://www.npmjs.com/package/editorconfig (which has 4+ million weekly downloads as of today) to parse the `.editorconfig`.

I didn't know where to put the function so the `lib/string.ts` seemed most appropriate.

Sadly I couldn't find a way to test it with a real `.editorconfig` file, so the tests only test if it fallbacks to `EOL` from `node:os` (If anybody is willing to give me a hint as to how to write the tests I will write them).

### CODE REVIEW

File: lib/string.ts [hunk 1]
```diff
@@ -24,11 +27,31 @@ export function capitalizeOnlyFirstLetter(str: string) {
...
 }
+
+// Gets the end of line string while respecting the
+// `.editorconfig` and falling back to `EOL` from `node:os`.
+export function getEndOfLine() {
+  // The passed string is the target file name, relative to process.cwd().
+  // Given that the docs markdown files are generated inside the `docs/rules` folder
+  // This should be the correct path to resolve the `.editorconfig` for these
+  // specific markdown files.
+  const config = editorconfig.parseSync('./docs/rules/markdown.md');
```
Comment (bmish) [inline lib/string.ts]: The path to the rule docs is configurable using [--path-rule-doc](https://github.com/bmish/eslint-doc-generator#configuration-options). For that matter, the path to the rule list(s) is configurable too with `--path-rule-list`.

Can we just lookup the `.editorconfig` for the plugin root instead? And not worry about nested doc folders.

File: test/lib/string-test.ts
Comment (bmish) [inline test/lib/string-test.ts:1]: I think we can add another test file. Try copying `test/lib/generate/sorting-test.ts` into a new test file called `test/lib/generate/editor-config-test.ts` and see if you can simulate `.editorconfig` using `mockFs` which should already be included in the test.

File: lib/string.ts (hunk 1 above)
Comment (CloudNStoyan) [reply lib/string.ts]: Sure, in hindsight, having an `.editorconfig` that has a different end_of_line setting only for specific markdown files in a specific folder is probably a really niche case.

File: test/lib/string-test.ts
Comment (CloudNStoyan) [reply test/lib/string-test.ts:1]: Much appreciated, I will give it a go tomorrow :)

**Comment by CloudNStoyan** [comment]
Using the `mockFs` worked the tests are now back at 100% coverage, thanks :)

File: test/lib/generate/editor-config-test.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
+import { getEndOfLine } from '../../../lib/string.js';
+import mockFs from 'mock-fs';
+import { jest } from '@jest/globals';
+
+describe('string (getEndOfLine)', function () {
```
Comment (bmish) [inline test/lib/generate/editor-config-test.ts:13]: Can you add a test in here with an empty `.editorconfig` and also a non-existent `.editorconfig`?

File: lib/string.ts [hunk 3]
```diff
@@ -39,11 +39,7 @@ export function sanitizeMarkdownTable(
 // Gets the end of line string while respecting the
 // `.editorconfig` and falling back to `EOL` from `node:os`.
 export function getEndOfLine() {
-  // The passed string is the target file name, relative to process.cwd().
-  // Given that the docs markdown files are generated inside the `docs/rules` folder
-  // This should be the correct path to resolve the `.editorconfig` for these
-  // specific markdown files.
-  const config = editorconfig.parseSync('./docs/rules/markdown.md');
+  const config = editorconfig.parseSync('markdown.md');
```
Comment (bmish) [inline lib/string.ts:45]: Can you add a comment about how we're just using a hypothetical markdown file in the plugin root to check for any markdown config?

File: test/lib/generate/editor-config-test.ts (hunk 2 above)
Comment (CloudNStoyan) [reply test/lib/generate/editor-config-test.ts:13]: Sure, do you want me to remove the test from `string-test.ts` named `handles when .editorconfig is not available and fallbacks to 'EOL' from 'node:os'` as it will overlap with the non-existent `.editorconfig` test?

Comment (bmish) [inline test/lib/generate/editor-config-test.ts:13]: Can you add at least one full test in here (copy from `test/lib/generate/sorting-test.ts`) that actually runs `generate()` and verifies that all the snapshots match? That way, we can test that `getEndOfLine` is actually being used everywhere and the snapshots end up with the correct character.

Comment (bmish) [reply test/lib/generate/editor-config-test.ts:13]: I think it's fine to have both since one is a unit test and one is an integration test.

**Comment by CloudNStoyan** [comment]
I had to move some of the `getEndOfLine` invocations inside their respective methods because the `mock-fs` couldn't mock them because the function was invoked when the generator function is imported instead of when it was ran.

Given that the `README.md` mock has newline in it, it works great to test if the rest of the code is using the same newline but as far as I can tell the Jest snapshot doesn't respect our newline choices.

**Comment by bmish** [comment]
No need to use `JSON.stringify`. As long as the snapshot test will fail if the wrong EOL character is used, then I'm happy with it.

**Review by bmish — APPROVED** [review]
Thanks a lot! Been wanting to get this bug fixed for a long time.
